### PR TITLE
Add macOS nanoCLR host target scaffold

### LINKED ISSUE #1708: Add macOS host nanoCLR target scaffold (targets/osx)
### Description

This issue proposes adding an initial macOS host target scaffold under:
```
targets/osx
```

The goal is to introduce a buildable and CI-validated entry point for a native macOS host nanoCLR implementation without impacting existing targets (Win32, RTOS boards, etc.).

This is intentionally a minimal scaffold:

- CMake + Ninja build entrypoint for macOS
- Host glue structure aligned with `targets/win32`
- Basic smoke executable (`--help`, `--version`)
- GitHub Actions workflow validating macOS build

No shared CLR runtime integration is included in this initial step.

This approach allows incremental and safe portability work without destabilizing current platforms.

---

## Motivation

- Enable native macOS CI validation.
- Provide a structured landing zone for future host nanoCLR work.
- Allow gradual portability improvements without modifying shared runtime behavior prematurely.

By introducing a scaffold first, runtime integration can be implemented incrementally in follow-up PRs.

---

## Non-goals (for this initial step)

- No managed filesystem support.
- No networking support.
- No integration with shared runtime startup (`src/CLR/**`).
- No functional parity with Win32 host target.

---

## Acceptance Criteria

- `targets/osx` folder exists with:
  - `CMakeLists.txt`
  - `README.md`
  - `Include/`
  - `nanoCLR/`
- macOS build works via CMake + Ninja (arm64).
- Smoke executable builds and runs (`--version` returns exit code 0).
- GitHub Actions workflow builds and runs smoke on `macos-14`.
- No changes to existing targets or shared runtime behavior.
- All new C/C++ files comply with project clang-format rules.

### How to solve the problem

Introduce a macOS host target scaffold under `targets/osx` with a CMake + Ninja build entrypoint and minimal host glue.

The initial implementation provides:
- A smoke executable for build validation.
- A macOS CI workflow.
- Basic host portability shims to isolate platform-specific code.

No shared CLR runtime wiring is included yet.  
Follow-up work will introduce a small host abstraction layer to gradually remove Windows-specific assumptions from VIRTUAL_DEVICE paths.

### Describe alternatives you've considered

_No response_

### Aditional context

_No response_

### PR DESCRIPTION
**Pull request by asp2286**
Add macOS nanoCLR host target scaffold

## Description

- Adds a new isolated macOS host target under `targets/osx`.
- Introduces a CMake + Ninja build entrypoint for macOS.
- Provides minimal host glue aligned with the existing `targets/win32` structure.
- Includes a smoke executable (`--help`, `--version`) for build validation.
- Adds a GitHub Actions workflow to build and run the smoke executable on `macos-14`.
- Ensures all new C/C++ files comply with project clang-format rules.
- Does not modify shared runtime behavior or existing targets.

## Motivation and Context

This change introduces a safe and incremental starting point for a native macOS host nanoCLR implementation.

It enables:

- Native macOS CI validation.
- A structured landing zone for host portability work.
- Incremental introduction of host abstraction and portability shims without destabilizing existing targets.

Resolves nanoFramework/Home#1708

## How Has This Been Tested?

- Built locally on macOS (arm64) using CMake + Ninja.
- Verified smoke executable runs and returns exit code 0 with `--version`.
- Verified GitHub Actions workflow builds successfully on `macos-14`.
- Confirmed clang-format compliance using project formatting rules.

## Types of changes

- [x] New feature (non-breaking change which adds functionality to code)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies/declarations (update dependencies or assembly declarations and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the CONTRIBUTING document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

## Summary by CodeRabbit

* **New Features**
  * macOS host target for nanoCLR (arm64 & x86_64): lightweight launcher with smoke-mode, CLI (help/version/load), and host services (timing, events, memory heaps, file I/O, randomness, watchdog and minimal HAL/startup support).

* **Chores**
  * Platform scaffolding and CMake integration, README documenting build/config/limitations, and CI updates (GitHub Actions workflow and pipeline entries to build and publish macOS artifacts).

### CODE REVIEW

File: targets/osx/nanoCLR/targetPAL_Events.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        return;
+    }
+
+    *timerCompleteFlag = false;
+
+    std::thread([timerCompleteFlag, millisecondsFromNow]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(millisecondsFromNow));
+        *timerCompleteFlag = true;
+    }).detach();
+}
```
[the remark this answers is not in the document]
Comment (asp2286) [reply targets/osx/nanoCLR/targetPAL_Events.cpp:53]: Thanks for the review.

This implementation mirrors the current Win32 host target behavior, where the timer callback also writes directly to a `bool*` without additional synchronization.

At this stage the macOS target is a minimal host scaffold focused on build and behavioral parity (single active timer + cancel previous). Introducing a `reinterpret_cast<std::atomic<bool>*>` would assume the pointed object is actually an `std::atomic<bool>`, which is not guaranteed by the existing HAL contract and could introduce undefined behavior.

In a follow-up PR, when wiring the target into a proper completion/event loop, this will be refactored to use a queue-based completion path instead of writing to the flag from a background thread.

For now, behavior is aligned with existing host targets.

**Comment by asp2286** [comment]
@coderabbitai full review

File: targets/osx/Include/nanoHAL_Capabilites.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+//
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+#ifndef NANOHAL_CAPABILITIES_H
+#define NANOHAL_CAPABILITIES_H
+
+#endif // NANOHAL_CAPABILITIES_H
```
[the remark this answers is not in the document]
Comment (asp2286) [reply targets/osx/Include/nanoHAL_Capabilites.h:9]: Thanks for spotting this.

The misspelling appears to be legacy and exists across the repository, not introduced by this PR. Renaming these files would be a cross-cutting change affecting multiple targets and HAL sources.

To keep this PR focused on the macOS host scaffold, I suggest addressing the filename normalization in a dedicated cleanup PR.

Happy to open a follow-up housekeeping PR for this.

**Comment by asp2286** [comment]
> /azp run

@Ellerbach thanks again for triggering the Azure pipeline.

I was able to reproduce the failure locally.  
The initial build was forcing `-DNANO_OSX_ARCH=arm64`, which causes a `Bad CPU type in executable` error on Intel macOS agents.

I validated the fix on:
- Apple Silicon (arm64) – builds and runs successfully.
- Intel macOS (x86_64) – reproduced the failure and confirmed the dynamic arch detection resolves it.

The pipeline now detects the agent architecture via `uname -m` and configures `NANO_OSX_ARCH` accordingly.

**Comment by josesimoes** [comment]
Hi @asp2286,

Appreciate your work on putting up a MacOS target. As you probably know we have the nanoCLR distributed as [nanoclr](https://www.nuget.org/packages/nanoclr) .NET tool. Have you tested this to run on MacOS?

I'm asking because that's the target we've been using for quite some time to run a virtual device and that is used by the test framework too. There is an old WIN32 target of which the new nanoclr shares files with. The C# wrapper interacts with it through Interop.

On a quick look at the PR, I got the impression that you're mixing old (like command parsing) and the new architecture.

If the existing nanoclr tool does not run on MacOS, I would rather have "fixed" whatever is needed in the existing project(s) instead of creating another target.

Don't want to sound like I'm dismissing your work and effort. Just looking at this from a broader perspective. That's one of the reasons we have in the contributor guideline a note asking developers to discuss with the core team breaking or big workloads before committing to an approach. No one likes to waste time and effort! 😉

**Comment by josesimoes** [comment]
From your comments it's clear that you've looked deep into the code base.
Yes, the convergence to a single target netcore would be the ultimate goal for this. In the process, the work done would also take us closer to have *nix builds.

A couple of thoughts: 
1. The reason for using MSVC is mostly historical. The original port was an emulator and linked directly to C/C++ code files from the CLR. As the project was already there, it felt easier and smooth to "just" keep working and evolving it. At some point, in an effort to modernize it and open up the possibility to run this on other platforms and allow other Wire Protocol transport layers, the .NET CLI was introduced. The test framework started using it and, to be honest, the WIN32 solution only remains there by lack of time to do the final migration of the code files to netcore and get rid of it completely.
2. As all the other targets that run on an MCU, CMake is the build system. With VS fully supporting it, it can easily be integrated in the .NET solution and move forward with this.
3. Understood your notes, comments and reasoning so I suppose we're reasonably aligned by now.

To wrap up this: let's make targets/netcore backend cross-platform (CMake + portable shims) and ship per-RID native assets.

File: targets/osx/nanoCLR/ParseOptions_OSX.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+struct NanoClrHostOptions
+{
+    bool showHelp = false;
+    bool showVersion = false;
+    std::vector<std::string> assemblyPaths;
+};
+
+// TODO: replace with CLR_RT_ParseOptions integration once shared runtime code is enabled.
+bool nanoCLR_OSX_ParseOptions(
```
Comment (josesimoes) [inline targets/osx/nanoCLR/ParseOptions_OSX.cpp]: The options parser was moved to the C# wrapper.

**Review by josesimoes — COMMENTED** [review]
Looking good.
I need to test the part related with HHD interop DLL loading. I recall that one is very picky...

Apart from that, looking good.

**Review by josesimoes — APPROVED** [review]
Looks good to me.
Want to address any of Copilot review points before merging?

**Comment by Ellerbach** [comment]
@asp2286 you are right, this is needed. Now, what I'm figuring out as I started to have a deeper look at this as I was implementing the tests in the VS Code extension is that I can probably generalize this work to have a POSIX version of it which will then properly work on Linux and MacOS.

I'll work on this for couple of days and will report here and in a branch the work done and we can see how best move this forward.

Thanks again for the work, it's definitely a super solid foundation and sorry if we missed the point before!

**Comment by Ellerbach** [comment]
@asp2286 look at #3313 
the work to get that Mac OS virtual device and also for Linux is done! If you want to test, you can download those from the artifact! And feedback welcome.
So, I'll go and close this PR. It's been really useful to boostrap the work and understand what was missing! Thanks again!
